### ISSUE TITLE
allowAutomaticSingleRunInference causes incorrect type information

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#6184]
## Overview

Forking from #6176:

> It's a bit funky with --fix from what I've noticed.
> The fix re-runs won't have proper type information so rules like the no-unsafe-* rules will report false errors on fixed files that won't show up on a non-fixing run.
>
> We need to figure out a better solution for the fix case before we can really roll this out to everyone I think, or else we'll create a really flaky and confusing experience.

@bradzacher - do you have a repro available? I can work on one if not.

I'd consider this a bug we should tackle. Perhaps we can use the same builder/watch programs that we do for IDEs?

**Comment by bradzacher** [comment]
I can't seem to repro it right now - but I do remember seeing it on our repo. I did a `--fix` run and I got a bunch of `no-unsafe-*` errors which ofc disappeared after a second run.

There's probably a complex combination of library types or something that is needed.
I tried creating an isolated repro here https://github.com/bradzacher/6184-repro

**Comment by bradzacher** [comment]
There is one issue here: https://github.com/typescript-eslint/typescript-eslint/issues/3851
IDK if it's still an issue (needs re-investigation)

**Comment by JoshuaKGoldberg** [comment]
#3851 still repros for me. Marking this issue as a duplicate of that one. "Great"!
